### DISCUSSION TITLE
Constraints formulas in the PF module

### DISCUSSION THREAD

**Discussion by LiuPengPeter** [idaholab/moose#16085] [Q&A Modules: Phase field] [answered]
Hello,

as you know, there are two constraints formulas in PF:
eta1 + eta2 +eta3 =1; and eta1 * c1 + eta2 * c2 +eta3 * c3 =c.

**In ICs**, I write some codes **let eta3 equals 1- eta1 - eta2. And c equals ... in the whole area**. And **there doesn't need any constraint in Kernels.**

Right?

**Comment by WilkAndy** [comment]
Does "PF" mean "PhaseField" or "PorousFlow"?

**Reply by LiuPengPeter** [reply]
Sorry.
That's phase field.

**Comment by LiuPengPeter** [comment, marked as answer]
I just wrote codes about eta1 * c1 + eta2 * c2 +eta3 * c3 =c, and there is no error when making.
But I am afraid there will be problems. Could anyone help me check the expressions of computeQpJacobian and computeQpOffDiagJacobian? 
Thanks a lot!

Real
ConstraintConcentration::computeQpResidual()
{
  Real sum = 0.0;
  for (unsigned int i = 0; i < _cs_num; ++i)
    sum += (* _cs[i])[_qp] * (* _etas[i])[_qp];
  return _test[_i][_qp] * (sum - _u[_qp]);
}

Real
ConstraintConcentration::computeQpJacobian()
{
  return -_test[_i][_qp] * _phi[_j][_qp];
}

Real
ConstraintConcentration::computeQpOffDiagJacobian(unsigned int jvar)
{
  auto cjvar = mapJvarToCvar(jvar, _cs_map);
  if (cjvar >= 0)
    return _test[_i][_qp] * (* _etas[cjvar])[_qp] * _phi[_j][_qp];

  auto etavar = mapJvarToCvar(jvar, _etas_map);
  if (etavar >= 0)
  {
    return _test[_i][_qp] * (* _cs[etavar])[_qp] * _phi[_j][_qp];
  }

  return 0.0;
}

**Reply by dschwen** [reply]
That looks correct if you are applying that kernel to _c_. Correct use of `mapJvarToCvar`. Nice.

**Reply by LiuPengPeter** [reply]
Thanks for your encouragement!
